### PR TITLE
FEATURE: topic_view_stats table with daily fidelity

### PR DESCRIPTION
**Pull request by SamSaffron**
This gives us daily fidelity of topic view stats

New table stores a row per topic viewed per day tracking
anonymous and logged on views

### CODE REVIEW

File: app/models/topic_view_item.rb [hunk 1]
```diff
@@ -43,15 +44,30 @@ def self.add(topic_id, ip, user_id = nil, at = nil, skip_redis = false)
...
         if result > 0
           if user_id
             UserStat.where(user_id: user_id).update_all "topics_entered = topics_entered + 1"
           end
         end
 
-        # Update the views count in the parent, if it exists.
+        Topic.where(id: topic_id).update_all "views = views + 1"
+
+        sql = <<~SQL
```
Comment (martin-brennan) [inline app/models/topic_view_item.rb:55]: This is pretty minor but it feels a bit strange to have this code here but I get why, so we can keep the `TopicViewItem.add` interface unchanged. Maybe we could add a `TopicViewStat.add`, move this code there, and have:

```ruby
Topic.where(id: topic_id).update_all "views = views + 1"
TopicViewStat.add(topic_id: topic_id, viewed_at: at, user_id: user_id)
```

This at least indicates another table is involved now.

File: spec/models/topic_view_item_spec.rb [hunk 2]
```diff
@@ -6,6 +6,33 @@ def add(topic_id, ip, user_id = nil)
     TopicViewItem.add(topic_id, ip, user_id, nil, skip_redis)
   end
 
+  it "correctly increase topic view stats" do
```
Comment (martin-brennan) [inline spec/models/topic_view_item_spec.rb:9]: I think it's worth adding some request specs here too:

https://github.com/discourse/discourse/blob/dd2fa58dd187a1298c56536a98dbd13ca11c4012/spec/requests/topics_controller_spec.rb#L2700-L2702

Just to ensure that both anon and logged in user views are logged correctly from that level

**Review by martin-brennan — APPROVED** [review]
Looks good to me, only some minor comments. You already addressed thoughts around retention etc. internally. The only other thing we may want to add is a default on `enable_detailed_topic_view_logging` setting to hide this behind, but also not strictly necessary.

File: spec/models/topic_view_item_spec.rb (hunk 2 above)
Comment (SamSaffron) [reply spec/models/topic_view_item_spec.rb:9]: this is going to break when we amend "how we count", we will be forced to fix the test to add a header, might as well just do the testing then.

File: app/models/topic_view_item.rb (hunk 1 above)
Comment (SamSaffron) [reply app/models/topic_view_item.rb]: done

**Pull request merged**